BUG: Building with vs2019 fails in SIMD dispatch code

Building latest HEAD with Visual Studio 14.22.27905 (Visual Studio 2019 Community Edition) fails, there are warnings that are turned into an error:
```
_simd.dispatch.avx512_skx.c
numpy\core\src\_simd\_simd.dispatch.c.src(425):
warning C4244: 'initializing': conversion from '__mmask16' to 'npyv_b64', \
      possible loss of data
... and 3 more like that
```

The warnings point to this code (the first SIMD_IMPL_INTRIN_2` is line 425):
```
// Logical
/**begin repeat
 * #bsfx = b8, b16, b32, b64#
 */
SIMD_IMPL_INTRIN_2(and_@bsfx@, v@bsfx@, v@bsfx@, v@bsfx@)
SIMD_IMPL_INTRIN_2(or_@bsfx@,  v@bsfx@, v@bsfx@, v@bsfx@)
SIMD_IMPL_INTRIN_2(xor_@bsfx@, v@bsfx@, v@bsfx@, v@bsfx@)
SIMD_IMPL_INTRIN_1(not_@bsfx@, v@bsfx@, v@bsfx@)
```
It seems the first line is preprocessed into (for `b16`):
```
#line 425 "numpy\\core\\src\\_simd\\_simd.dispatch.c.src"
static PyObject *simd__intrin_and_b16 (PyObject* (__NPY_UNUSED_TAGGEDself) , PyObject *args) {
    simd_arg arg1 = {.dtype = simd_data_vb16}; 
    simd_arg arg2 = {.dtype = simd_data_vb16};
    if (!PyArg_ParseTuple( args, "O&O&:""and_b16", simd_arg_converter, &arg1, simd_arg_converter, &arg2 ))
        return ((void *)0); 
    simd_data data = {.vb16 = _kand_mask32( arg1.data.vb16, arg2.data.vb16 )};
    simd_arg_free(&arg1); 
    simd_arg_free(&arg2); 
    simd_arg ret = { .data = data, .dtype = simd_data_vb16 };
    return simd_arg_to_obj(&ret); 
}
```
Note the call to `_kand_mask32` in the middle. That is defined elsewhere as
```
extern __mmask32 __cdecl _kand_mask32(__mmask32, __mmask32);
```

So somehow the macro is setting up the wrong function?

<details>
<summary>`build.log`</summary>

Some of the log lines are very long.

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -Ibuild\src.win-amd64-3.8\numpy\core\src\_simd -Inumpy\core\include -Ibuild\src.win-amd64-3.8\numpy\core\include/numpy -Ibuild\src.win-amd64-3.8\numpy\distutils\include -Inumpy\core\src\common -Inumpy\core\src -Inumpy\core -Inumpy\core\src\npymath -Inumpy\core\src\multiarray -Inumpy\core\src\umath -Inumpy\core\src\npysort -Inumpy\core\src\_simd -IC:\pypy_stuff\Python38\include -IC:\pypy_stuff\Python38\include -Ibuild\src.win-amd64-3.8\numpy\core\src\common -Ibuild\src.win-amd64-3.8\numpy\core\src\npymath -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\ATLMFC\include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt -IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE -IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt /Tcbuild\src.win-amd64-3.8\numpy\core\src\_simd\_simd.dispatch.avx512_skx.c /Fobuild\temp.win-amd64-3.8\Release\build\src.win-amd64-3.8\numpy\core\src\_simd\_simd.dispatch.avx512_skx.obj /WX /arch:AVX512
Running from numpy source directory.
C:\pypy_stuff\numpy\numpy\distutils\system_info.py:1989: UserWarning: 
    Optimized (vendor) Blas libraries are not found.
    Falls back to netlib Blas library which has worse performance.
    A better performance should be easily gained by switching
    Blas library.
  if self._calc_info(blas):
C:\pypy_stuff\numpy\numpy\distutils\system_info.py:1989: UserWarning: 
    Blas (http://www.netlib.org/blas/) libraries not found.
    Directories to search for the libraries can be specified in the
    numpy/distutils/site.cfg file (section [blas]) or by setting
    the BLAS environment variable.
  if self._calc_info(blas):
C:\pypy_stuff\numpy\numpy\distutils\system_info.py:1989: UserWarning: 
    Blas (http://www.netlib.org/blas/) sources not found.
    Directories to search for the sources can be specified in the
    numpy/distutils/site.cfg file (section [blas_src]) or by setting
    the BLAS_SRC environment variable.
  if self._calc_info(blas):
C:\pypy_stuff\numpy\numpy\distutils\system_info.py:1849: UserWarning: 
    Lapack (http://www.netlib.org/lapack/) libraries not found.
    Directories to search for the libraries can be specified in the
    numpy/distutils/site.cfg file (section [lapack]) or by setting
    the LAPACK environment variable.
  return getattr(self, '_calc_info_{}'.format(name))()
C:\pypy_stuff\numpy\numpy\distutils\system_info.py:1849: UserWarning: 
    Lapack (http://www.netlib.org/lapack/) sources not found.
    Directories to search for the sources can be specified in the
    numpy/distutils/site.cfg file (section [lapack_src]) or by setting
    the LAPACK_SRC environment variable.
  return getattr(self, '_calc_info_{}'.format(name))()
Warning: attempted relative import with no known parent package
C:\pypy_stuff\Python38\lib\distutils\dist.py:274: UserWarning: Unknown distribution option: 'define_macros'
  warnings.warn(msg)
_simd.dispatch.avx512_skx.c
numpy\core\src\_simd\_simd.dispatch.c.src(425): error C2220: warning treated as error - no 'object' file generated
numpy\core\src\_simd\_simd.dispatch.c.src(425): warning C4244: 'initializing': conversion from '__mmask16' to 'npyv_b64', possible loss of data
numpy\core\src\_simd\_simd.dispatch.c.src(426): warning C4244: 'initializing': conversion from '__mmask16' to 'npyv_b64', possible loss of data
numpy\core\src\_simd\_simd.dispatch.c.src(427): warning C4244: 'initializing': conversion from '__mmask16' to 'npyv_b64', possible loss of data
numpy\core\src\_simd\_simd.dispatch.c.src(428): warning C4244: 'initializing': conversion from '__mmask16' to 'npyv_b64', possible loss of data

########### EXT COMPILER OPTIMIZATION ###########
Platform      : 
  Architecture: x64
  Compiler    : msvc

CPU baseline  : 
  Requested   : 'min'
  Enabled     : SSE SSE2 SSE3
  Flags       : none
  Extra checks: none

CPU dispatch  : 
  Requested   : 'max -xop -fma4'
  Enabled     : SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   : 
              : 
  SSE41       : SSE SSE2 SSE3 SSSE3
  Flags       : 
  Extra checks: none
  Detect      : SSE SSE2 SSE3 SSSE3 SSE41
              : numpy\core\src\umath\_umath_tests.dispatch.c
              : 
  SSE42       : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT
  Flags       : 
  Extra checks: none
  Detect      : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42
              : build\src.win-amd64-3.8\numpy\core\src\_simd\_simd.dispatch.c
              : 
  AVX2        : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3
  Flags       : /arch:AVX2
  Extra checks: none
  Detect      : AVX F16C FMA3 AVX2
              : build\src.win-amd64-3.8\numpy\core\src\umath\loops_arithm_fp.dispatch.c
              : build\src.win-amd64-3.8\numpy\core\src\umath\loops_trigonometric.dispatch.c
              : numpy\core\src\umath\_umath_tests.dispatch.c
              : build\src.win-amd64-3.8\numpy\core\src\_simd\_simd.dispatch.c
              : 
  AVX512F     : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512CD AVX512_SKX
  Flags       : /arch:AVX512
  Extra checks: none
  Detect      : AVX512_SKX
              : build\src.win-amd64-3.8\numpy\core\src\umath\loops_arithm_fp.dispatch.c
              : build\src.win-amd64-3.8\numpy\core\src\umath\loops_trigonometric.dispatch.c
              : 
  AVX512_SKX  : SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD
  Flags       : /arch:AVX512
  Extra checks: AVX512BW_MASK
  Detect      : AVX512_SKX
              : build\src.win-amd64-3.8\numpy\core\src\_simd\_simd.dispatch.c
CCompilerOpt._cache_write[796] : write cache to path -> C:\pypy_stuff\numpy\build\temp.win-amd64-3.8\Release\ccompiler_opt_cache_ext.py

########### CLIB COMPILER OPTIMIZATION ###########
Platform      : 
  Architecture: x64
  Compiler    : msvc

CPU baseline  : 
  Requested   : 'min'
  Enabled     : SSE SSE2 SSE3
  Flags       : none
  Extra checks: none

CPU dispatch  : 
  Requested   : 'max -xop -fma4'
  Enabled     : SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   : none
CCompilerOpt._cache_write[796] : write cache to path -> C:\pypy_stuff\numpy\build\temp.win-amd64-3.8\ccompiler_opt_cache_clib.py
error: Command "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -Ibuild\src.win-amd64-3.8\numpy\core\src\_simd -Inumpy\core\include -Ibuild\src.win-amd64-3.8\numpy\core\include/numpy -Ibuild\src.win-amd64-3.8\numpy\distutils\include -Inumpy\core\src\common -Inumpy\core\src -Inumpy\core -Inumpy\core\src\npymath -Inumpy\core\src\multiarray -Inumpy\core\src\umath -Inumpy\core\src\npysort -Inumpy\core\src\_simd -IC:\pypy_stuff\Python38\include -IC:\pypy_stuff\Python38\include -Ibuild\src.win-amd64-3.8\numpy\core\src\common -Ibuild\src.win-amd64-3.8\numpy\core\src\npymath -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\ATLMFC\include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt -IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE -IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt /Tcbuild\src.win-amd64-3.8\numpy\core\src\_simd\_simd.dispatch.avx512_skx.c /Fobuild\temp.win-amd64-3.8\Release\build\src.win-amd64-3.8\numpy\core\src\_simd\_simd.dispatch.avx512_skx.obj /WX /arch:AVX512" failed with exit status 2
```

--- a/numpy/core/src/common/simd/avx512/conversion.h
+++ b/numpy/core/src/common/simd/avx512/conversion.h
@@ -119,7 +119,13 @@ NPY_FINLINE npy_uint64 npyv_tobits_b16(npyv_b16 a)
 NPY_FINLINE npy_uint64 npyv_tobits_b32(npyv_b32 a)
 { return (npy_uint16)a; }
 NPY_FINLINE npy_uint64 npyv_tobits_b64(npyv_b64 a)
-{ return (npy_uint8)a; }
+{
+#ifdef NPY_HAVE_AVX512DQ_MASK
+    return _cvtmask8_u32(a);
+#else
+    return (npy_uint8)a;
+#endif
+}
 
 // round to nearest integer (assuming even)
 #define npyv_round_s32_f32 _mm512_cvtps_epi32

--- a/numpy/core/src/common/simd/avx512/operators.h
+++ b/numpy/core/src/common/simd/avx512/operators.h
@@ -90,21 +90,6 @@
     NPYV_IMPL_AVX512_FROM_SI512_PS_2ARG(npyv_and_f32, _mm512_and_si512)
     NPYV_IMPL_AVX512_FROM_SI512_PD_2ARG(npyv_and_f64, _mm512_and_si512)
 #endif
-#ifdef NPY_HAVE_AVX512BW_MASK
-    #define npyv_and_b8  _kand_mask64
-    #define npyv_and_b16 _kand_mask32
-#elif defined(NPY_HAVE_AVX512BW)
-    NPY_FINLINE npyv_b8 npyv_and_b8(npyv_b8 a, npyv_b8 b)
-    { return a & b; }
-    NPY_FINLINE npyv_b16 npyv_and_b16(npyv_b16 a, npyv_b16 b)
-    { return a & b; }
-#else
-    #define npyv_and_b8  _mm512_and_si512
-    #define npyv_and_b16 _mm512_and_si512
-#endif
-#define npyv_and_b32 _mm512_kand
-#define npyv_and_b64 _mm512_kand
-
 // OR
 #define npyv_or_u8  _mm512_or_si512
 #define npyv_or_s8  _mm512_or_si512
@@ -121,20 +106,6 @@
     NPYV_IMPL_AVX512_FROM_SI512_PS_2ARG(npyv_or_f32, _mm512_or_si512)
     NPYV_IMPL_AVX512_FROM_SI512_PD_2ARG(npyv_or_f64, _mm512_or_si512)
 #endif
-#ifdef NPY_HAVE_AVX512BW_MASK
-    #define npyv_or_b8  _kor_mask64
-    #define npyv_or_b16 _kor_mask32
-#elif defined(NPY_HAVE_AVX512BW)
-    NPY_FINLINE npyv_b8 npyv_or_b8(npyv_b8 a, npyv_b8 b)
-    { return a | b; }
-    NPY_FINLINE npyv_b16 npyv_or_b16(npyv_b16 a, npyv_b16 b)
-    { return a | b; }
-#else
-    #define npyv_or_b8  _mm512_or_si512
-    #define npyv_or_b16 _mm512_or_si512
-#endif
-#define npyv_or_b32 _mm512_kor
-#define npyv_or_b64 _mm512_kor
 
 // XOR
 #define npyv_xor_u8  _mm512_xor_si512
@@ -152,21 +123,6 @@
     NPYV_IMPL_AVX512_FROM_SI512_PS_2ARG(npyv_xor_f32, _mm512_xor_si512)
     NPYV_IMPL_AVX512_FROM_SI512_PD_2ARG(npyv_xor_f64, _mm512_xor_si512)
 #endif
-#ifdef NPY_HAVE_AVX512BW_MASK
-    #define npyv_xor_b8  _kxor_mask64
-    #define npyv_xor_b16 _kxor_mask32
-#elif defined(NPY_HAVE_AVX512BW)
-    NPY_FINLINE npyv_b8 npyv_xor_b8(npyv_b8 a, npyv_b8 b)
-    { return a ^ b; }
-    NPY_FINLINE npyv_b16 npyv_xor_b16(npyv_b16 a, npyv_b16 b)
-    { return a ^ b; }
-#else
-    #define npyv_xor_b8  _mm512_xor_si512
-    #define npyv_xor_b16 _mm512_xor_si512
-#endif
-#define npyv_xor_b32 _mm512_kxor
-#define npyv_xor_b64 _mm512_kxor
-
 // NOT
 #define npyv_not_u8(A) _mm512_xor_si512(A, _mm512_set1_epi32(-1))
 #define npyv_not_s8  npyv_not_u8
@@ -183,21 +139,67 @@
     #define npyv_not_f32(A) _mm512_castsi512_ps(npyv_not_u32(_mm512_castps_si512(A)))
     #define npyv_not_f64(A) _mm512_castsi512_pd(npyv_not_u64(_mm512_castpd_si512(A)))
 #endif
+
+/***************************
+ * Logical (boolean)
+ ***************************/
 #ifdef NPY_HAVE_AVX512BW_MASK
+    #define npyv_and_b8  _kand_mask64
+    #define npyv_and_b16 _kand_mask32
+    #define npyv_or_b8   _kor_mask64
+    #define npyv_or_b16  _kor_mask32
+    #define npyv_xor_b8  _kxor_mask64
+    #define npyv_xor_b16 _kxor_mask32
     #define npyv_not_b8  _knot_mask64
     #define npyv_not_b16 _knot_mask32
 #elif defined(NPY_HAVE_AVX512BW)
-    NPY_FINLINE npyv_b8 npyv_not_b8(npyv_b8 a)
+    NPY_FINLINE npyv_b8  npyv_and_b8(npyv_b8 a, npyv_b8 b)
+    { return a & b; }
+    NPY_FINLINE npyv_b16 npyv_and_b16(npyv_b16 a, npyv_b16 b)
+    { return a & b; }
+    NPY_FINLINE npyv_b8  npyv_or_b8(npyv_b8 a, npyv_b8 b)
+    { return a | b; }
+    NPY_FINLINE npyv_b16 npyv_or_b16(npyv_b16 a, npyv_b16 b)
+    { return a | b; }
+    NPY_FINLINE npyv_b8  npyv_xor_b8(npyv_b8 a, npyv_b8 b)
+    { return a ^ b; }
+    NPY_FINLINE npyv_b16 npyv_xor_b16(npyv_b16 a, npyv_b16 b)
+    { return a ^ b; }
+    NPY_FINLINE npyv_b8  npyv_not_b8(npyv_b8 a)
     { return ~a; }
     NPY_FINLINE npyv_b16 npyv_not_b16(npyv_b16 a)
     { return ~a; }
 #else
+    #define npyv_and_b8  _mm512_and_si512
+    #define npyv_and_b16 _mm512_and_si512
+    #define npyv_or_b8   _mm512_or_si512
+    #define npyv_or_b16  _mm512_or_si512
+    #define npyv_xor_b8  _mm512_xor_si512
+    #define npyv_xor_b16 _mm512_xor_si512
     #define npyv_not_b8  npyv_not_u8
     #define npyv_not_b16 npyv_not_u8
 #endif
-#define npyv_not_b32 _mm512_knot
-#define npyv_not_b64 _mm512_knot
 
+#define npyv_and_b32 _mm512_kand
+#define npyv_or_b32  _mm512_kor
+#define npyv_xor_b32 _mm512_kxor
+#define npyv_not_b32 _mm512_knot
+
+#ifdef NPY_HAVE_AVX512DQ_MASK
+    #define npyv_and_b64 _kand_mask8
+    #define npyv_or_b64  _kor_mask8
+    #define npyv_xor_b64 _kxor_mask8
+    #define npyv_not_b64 _knot_mask8
+#else
+    NPY_FINLINE npyv_b64 npyv_and_b64(npyv_b64 a, npyv_b64 b)
+    { return (npyv_b64)_mm512_kand((npyv_b32)a, (npyv_b32)b); }
+    NPY_FINLINE npyv_b64 npyv_or_b64(npyv_b64 a, npyv_b64 b)
+    { return (npyv_b64)_mm512_kor((npyv_b32)a, (npyv_b32)b); }
+    NPY_FINLINE npyv_b64 npyv_xor_b64(npyv_b64 a, npyv_b64 b)
+    { return (npyv_b64)_mm512_kxor((npyv_b32)a, (npyv_b32)b); }
+    NPY_FINLINE npyv_b64 npyv_not_b64(npyv_b64 a)
+    { return (npyv_b64)_mm512_knot((npyv_b32)a); }
+#endif
 
 /***************************
  * Comparison

--- a/numpy/distutils/ccompiler_opt.py
+++ b/numpy/distutils/ccompiler_opt.py
@@ -259,7 +259,7 @@ class _Config:
         AVX512_SKX = dict(
             interest=42, implies="AVX512CD", group="AVX512VL AVX512BW AVX512DQ",
             detect="AVX512_SKX", implies_detect=False,
-            extra_checks="AVX512BW_MASK"
+            extra_checks="AVX512BW_MASK AVX512DQ_MASK"
         ),
         AVX512_CLX = dict(
             interest=43, implies="AVX512_SKX", group="AVX512VNNI",

--- a/numpy/distutils/checks/extra_avx512dq_mask.c
+++ b/numpy/distutils/checks/extra_avx512dq_mask.c
@@ -1,0 +1,16 @@
+#include <immintrin.h>
+/**
+ * Test DQ mask operations due to:
+ *  - MSVC has supported it since vs2019 see,
+ *    https://developercommunity.visualstudio.com/content/problem/518298/missing-avx512bw-mask-intrinsics.html
+ *  - Clang >= v8.0
+ *  - GCC >= v7.1
+ */
+int main(void)
+{
+    __mmask8 m8 = _mm512_cmpeq_epi64_mask(_mm512_set1_epi64(1), _mm512_set1_epi64(1));
+    m8 = _kor_mask8(m8, m8);
+    m8 = _kxor_mask8(m8, m8);
+    m8 = _cvtu32_mask8(_cvtmask8_u32(m8));
+    return (int)_cvtmask8_u32(m8);
+}

